File: DotNetGotchas/CSharp/ES/ESPlatformDiff/ESLib/Comp.cs
// Comp.cs part of ESLib.dll
using System;
using System.EnterpriseServices;

namespace ESLib
{
	[Transaction(TransactionOption.Required), JustInTimeActivation]
	public class Comp : ServicedComponent
	{
		private int theKey;
		private int theVal;

		internal void init(int key)
		{
			ContextUtil.MyTransactionVote 
				= TransactionVote.Abort;

			theKey = key;
			theVal = key * 10;

			ContextUtil.MyTransactionVote 
				= TransactionVote.Commit;
		}

		public int GetValue()
		{
			return theVal;
		}

		public void SetValue(int val)
		{
			ContextUtil.MyTransactionVote 
				= TransactionVote.Abort;

			theVal = val;

			if (val < 0)
			{
				ContextUtil.DeactivateOnReturn = true;
				throw new ApplicationException(
						"Invalid value");
			}

			ContextUtil.MyTransactionVote 
				= TransactionVote.Commit;
		}
	}
}

File: DotNetGotchas/VBDN/ES/ESPlatformDiff/ESLib/Comp.vb
' Comp.vb part of ESLib.dll
Imports System.EnterpriseServices

<Transaction(TransactionOption.Required), JustInTimeActivation()> _
Public Class Comp
    Inherits ServicedComponent

    Private theKey As Integer
    Private theVal As Integer

    Friend Sub init(ByVal key As Integer)
        ContextUtil.MyTransactionVote = TransactionVote.Abort

        theKey = key
        theVal = key * 10

        ContextUtil.MyTransactionVote = TransactionVote.Commit
    End Sub


    Public Function GetValue() As Integer
        Return theVal
    End Function

    Public Sub SetValue(ByVal val As Integer)
        ContextUtil.MyTransactionVote = TransactionVote.Abort

        theVal = val

        If val < 0 Then
            ContextUtil.DeactivateOnReturn = True
            Throw New ApplicationException("Invalid value")
        End If

        ContextUtil.MyTransactionVote = TransactionVote.Commit
    End Sub
End Class
